src/NovelRT/Experimental/Input/Glfw/GlfwInputDevice.cpp

```diff
@@ -0,0 +1,133 @@
+
+    if (previousBufferResult == previousBuffer.end())
+    {
+        previousStateResult = KeyState::Idle;
+    }
+    else
+    {
+        previousStateResult = previousBufferResult->second.getCurrentState();
+    }
+
+    auto currentBufferResult = currentBuffer.find(code);
+    KeyStateFrameChangeLog changeLogObject{};
+
+    if (currentBufferResult == currentBuffer.end())
+    {
+        changeLogObject = KeyStateFrameChangeLog();
+    }
+    else
+    {
+        changeLogObject = currentBufferResult->second;
+    }
+
+    switch (state)
+    {
+        case KeyState::KeyDown:
+            if (previousStateResult == KeyState::KeyDown)
+            {
+                changeLogObject.pushNewState(KeyState::KeyDownHeld);
+            }
+            else if (previousStateResult != KeyState::KeyDownHeld)
+            {
+                changeLogObject.pushNewState(KeyState::KeyDown); // TODO: Is this actually gonna work lol
+            }
+            break;
+        case KeyState::KeyDownHeld:
+        case KeyState::KeyUp:
+            changeLogObject.pushNewState((previousStateResult == KeyState::KeyUp) ? KeyState::Idle : state);
+            break;
+        case KeyState::Idle:
+        default:
+            // do nothing. Seriously. These cases are only here because the Ubuntu build cried at me in CI for not
+            // having them suddenly.
+            break;
+    }
+}
+
+void GlfwInputDevice::acceptMouseButtonClickPush(int32_t button, int32_t action, Maths::GeoVector2F mousePosition)
+{
+    auto keyState = static_cast<KeyState>(action);
+    auto keyCode = static_cast<KeyCode>(button);
+    auto value = Maths::GeoVector4F(mousePosition).vec4Value() *
+                 glm::scale(glm::vec3(1920.0f / _screenSize.x, 1080.0f / _screenSize.y, 0.0f));
+
+    _cursorPosition = Maths::GeoVector2F(value.x, value.y);
+    KeyStateFrameChangeLog log{};
+
+    if (_keyStates.at(_currentBufferIndex).find(keyCode) != _keyStates.at(_currentBufferIndex).end())
+    {
+        log = _keyStates.at(_currentBufferIndex).at(keyCode);
+    }
+
+    log.pushNewState(keyState);
+    _keyStates.at(_currentBufferIndex).insert_or_assign(keyCode, log);
+}
+
+void GlfwInputDevice::acceptKeyboardInputBindingPush(int32_t key, int32_t action)
+{
+    auto keyState = static_cast<KeyState>(action);
+    auto keyCode = static_cast<KeyCode>(key);
+    KeyStateFrameChangeLog log{};
+
+    if (_keyStates.at(_currentBufferIndex).find(keyCode) != _keyStates.at(_currentBufferIndex).end())
+    {
+        log = _keyStates.at(_currentBufferIndex).at(keyCode);
+    }
+
+    log.pushNewState(keyState);
+    _keyStates.at(_currentBufferIndex).insert_or_assign(keyCode, log);
+}
+
+void GlfwInputDevice::consumePlayerInput()
+{
+    _currentBufferIndex = (_currentBufferIndex + 1) % INPUT_BUFFER_COUNT;
+    _keyStates.at(_currentBufferIndex).clear();
+    glfwPollEvents();
+    processKeyStates();
+    _previousBufferIndex = _currentBufferIndex;
+}
```

Comment:
I'm guessing this is just currently a port of things as they are?

Reply:
That's right, I mostly wanted to move what you already had over to the new format and make sure I have the setup right. After reading your above comments I have a better sense for how much you want to change as well. I just found a good article for the other kinds of glfw input here: https://www.glfw.org/docs/3.3/input_guide.html . I'll start to actually make changes now and see if its what you had in mind. What would you recommend as a second library/api to look at so I can keep different ones in my mind when making changes?